### DISCUSSION TITLE
Private registry works with npm but not with yarn when deploying on vercel

### DISCUSSION THREAD

**Discussion by agcty** [vercel/vercel#5418] [General] [answered]
We're using yarn to build the project and have a private github npm registry to host some of our packages.
In out vercel dashboard, we've correctly set the NPM_RC variable. When deploying our next.js project with npm, the private packages get fetched and everything works as expected. However, when we do the same thing with yarn, the deployment fails with a 401 unauthorized error code when trying to access the private registry.

Locally, both yarn and npm work with the same .npmrc file.

Local npm version: 6.14.8
Local yarn version: 1.22.10

Vercel seems to deploy with yarn 1.22.4, not sure if this is an issue.

Has anyone managed to deploy a project both yarn and a private registry?

**Comment by agcty** [comment]
So after reaching out to vercel support, they showed me this issue: [https://github.com/yarnpkg/yarn/issues/3093](https://github.com/yarnpkg/yarn/issues/3093)

You have to make your .npmrc look exactly like this:

always-auth=true
@zerolens-com:registry=https://npm.pkg.github.com/YOUR_USERNAME/
//npm.pkg.github.com/:_authToken=YOUR_TOKEN
//npm.pkg.github.com/YOUR_USERNAME/:_authToken=YOUR_TOKEN
registry=https://registry.npmjs.org

You somehow need always-auth=true and both npm.pkg.github.com lines

**Reply by nathanielpomianowski** [reply]
Does anyone have a solution for this when using a Gitlab private registry under a self-hosted gitlab?

**Reply by amyegan** [reply]
For private GitLab resources, you could [create a deploy token](https://docs.gitlab.com/ee/user/project/deploy_tokens/#create-a-deploy-token) and use that as part of the address (source: https://stackoverflow.com/a/27015550)

**Reply by nathanielpomianowski** [reply]
I found a solution for anyone using Gitlab private registry:
```
always-auth=true
@<GROUP_OR_USER_NAME>:registry=https://<PRIVATE_INSTANCE_NAME>/api/v4/packages/npm/
//<PRIVATE_INSTANCE_NAME>/api/v4/projects/:_authToken=<PROJECT_ID>/packages/npm/:_authToken=<GITLAB_TOKEN>
//<PRIVATE_INSTANCE_NAME>/api/v4/packages/npm/:_authToken=<GITLAB_TOKEN>
registry=https:registry.npmjs.org
```
i.e.
<GROUP_OR_USER_NAME> == foo (Either a group or your individual name. Keep this lowercase)
<PRIVATE_INSTANCE_NAME> == gitlab.example.com
<PROJECT_ID> == 42
<GITLAB_TOKEN> == jlfwjl$@l2jlwjf

**Comment by allan-almeida** [comment, marked as answer]
For new versions of Yarn, (I am using `2.4.0`), I was able to successfully deploy to Vercel by adding the following to my `.yarnrc.yml` file:

```
npmScopes:
  YOUR_SCOPE:
    npmRegistryServer: 'https://npm.pkg.github.com'
    npmAlwaysAuth: true
    npmAuthToken: ${GITHUB_PACKAGE_TOKEN}
```
    
as per https://yarnpkg.com/configuration/yarnrc#npmScopes.
    
I also added `GITHUB_PACKAGE_TOKEN` as an environment variable to my builds in Vercel.

**Reply by heygrady** [reply]
This worked for me but it's not very portable to do it this way. It was causing issues on my local environment as some of my local apps aren't running in the same context. However, I was inspired by this solution and discovered a better way for me.

https://yarnpkg.com/configuration/yarnrc/

> Finally, note that most settings can also be defined through environment variables[...]. To do this, just prefix the names and write them in snake case: YARN_CACHE_FOLDER will set the cache folder.

I added a `YARN_NPM_AUTH_TOKEN` environment variable in my vercel project settings and that seemed to work great.
